src/olmo_core/data/source_mixture.py

```diff
@@ -0,0 +1,307 @@
+
+        if not self.source_configs:
+            raise OLMoConfigurationError("source_configs must not be empty")
+
+        if (total := sum([source.target_ratio for source in self.source_configs])) != 1.0:
+            raise OLMoConfigurationError(f"target_ratios must sum to 1, got {total}")
+
+    def build(self) -> SourceMixtureDataset:
+        self.validate()
+        random.seed(self.seed)
+        available_tokens_by_source: Dict[str, int] = {}
+
+        print("---------------------------------------------------------")
+        print("Generating a source mixture from configurations:")
+        for source_config in self.source_configs:
+            pprint(source_config)
+
+        # Count the number of tokens available for each source
+        for source_config in self.source_configs:
+            log.info(f"Counting tokens for source: {source_config.source_name}")
+            available_tokens_by_source[source_config.source_name] = self._count_tokens_for_paths(
+                paths=source_config.paths, source=source_config.source_name
+            )
+
+        tokens_details_by_source: List[SourceTokenDetails] = []
+
+        # Calculate the number of tokens available and to include for each source
+        for source_config in self.source_configs:
+            num_for_source = available_tokens_by_source[source_config.source_name]
+            needed_for_source = int(self.max_tokens * source_config.target_ratio)
+            max_for_source = int(
+                (num_for_source * source_config.max_source_fraction)
+                * source_config.max_repetition_ratio
+            )
+
+            # Ensure that the max tokens for a source meet the target ratio requirement
+            if max_for_source < needed_for_source:
+                raise OLMoConfigurationError(
+                    f"Insufficient tokens for source: {source_config.source_name} @ target global ratio: {source_config.target_ratio} :: {max_for_source} < {needed_for_source}"
+                )
+
+            tokens_details_by_source.append(
+                SourceTokenDetails(
+                    config=source_config,
+                    population=num_for_source,
+                    num_selected=needed_for_source,
+                )
+            )
+
+        completed: List[SourceMixtureOutcome] = []
+        for source in tokens_details_by_source:
+            completed.append(
+                SourceMixtureOutcome(
+                    name=source.config.source_name,
+                    path_tokens=self.get_paths_and_tokens_for_source(
+                        source_config=source.config,
+                        token_details=source,
+                    ),
+                )
+            )
+
+        log.info("Outcome by source => ")
+        print(
+            tabulate.tabulate(
+                [item.for_table(self.max_tokens) for item in tokens_details_by_source],
+                headers="keys",
+                tablefmt="pretty",
+            ),
```

Comment:
You can create nice looking tables with `rich`, which is already a dependency. I would prefer not to add another dependency just for pretty printing some stuff.